### DISCUSSION TITLE
Ranges: Support for Specification via Length

### DISCUSSION THREAD

**Discussion by Saalvage** [dotnet/csharplang#6093] [Language Ideas]
Ranges currently only support being specified via two indices. I believe that it might be a good idea to a allow their specification via one start index and a length.

If you wanted to emulate this behavior you would either need to use a helper variable or duplicate the expression used for the start index, which can be both a visual as well as a performance concern in the case of the latter solution.

### Syntax
I hope that people with a better idea on language design will propose a better syntax. For illustration purposes I will be using `start..=>length`.

### Implementation
Equipping `System.Range` with a new constructor taking in an index and an integer should do the trick on the API side of things.
The end index can simply be calculated by the start index +/- (depending on whether the index is from the end) the length.

### Example
```cs
string[] GetPage(int index)
    => string.Join('\n', _lines[(index * _pageSize)..(index * _pageSize) + _pageSize]);

// v

string[] GetPage(int index)
    => string.Join('\n', _lines[(index * _pageSize)..=>_pageSize]);
```

**Comment by svick** [comment]
Being able to create a `Range` from a start index and length sounds useful, but I'm not convinced it needs dedicated C# syntax, especially since it's not clear what syntax should be used. Instead, adding just the API to `Range` and using that directly sounds like a better approach to me.

> Equipping `System.Range` with a new constructor taking in an index and an integer should do the trick on the API side of things.

Note that this would be a breaking change. Today, code like `new Range(1, 2)` compiles because of the implicit conversion from `int` to `Index`. But adding the suggested constructor would change the meaning of that code.

Adding this API in a different way (probably as a `static` method) should be fine though.

**Reply by jnm2** [reply]
> Today, code like `new Range(1, 2)` compiles because of the implicit conversion from `int` to `Index`. But adding the suggested constructor would change the meaning of that code.

I'm not seeing how. Can you explain what the change in meaning is?

**Reply by Saalvage** [reply]
> > Today, code like `new Range(1, 2)` compiles because of the implicit conversion from `int` to `Index`. But adding the suggested constructor would change the meaning of that code.
> 
> I'm not seeing how. Can you explain what the change in meaning is?

Right now `new Range(1, 2)` constructs a range [1, 2), while with my originally proposed change it would result in a range [1, 3) since the 2 is no longer considered the numerical end index, but the length of the range.

I was unaware of the implicit conversion of `int` to `Index`, which is why I failed to consider it in my original post.

**Reply by jnm2** [reply]
Oh of course, thank you. I was thinking of the Roslyn type TextSpan rather than Range and thinking Range had an `int start, int length` constructor. Sorry about that.
